### DISCUSSION TITLE
Can we use react-router-dom in Next.js and still achieve server side rendering?

### DISCUSSION THREAD

**Discussion by annac-117** [vercel/next.js#50633] [Help] [answered]
### Summary

I have quite an unusual request from my client, they want to migrate to the nextjs app from CRA, but they don't want to use the nextjs inbuilt file based routing system but use react-router-dom , since they don't want to replace too much code during migration.

I've done what he asked for by using react-router-dom in the nextjs app, and now although the files inside pages folder are being rendered, they're being rendered client side (which is obvious). Now he wants me to fetch props into these pages files by using getServerSideProps! But since now the files inside pages folder are client side rendered, the props from getServerSideProps are showing as undefined on the client side.

So, is there any way to achieve server side rendering in nextjs pages folder using react-router-dom?

I know I'm going against the best practices here, but this is what the client wants.

**Comment by devjiwonchoi** [comment, marked as answer]
First of all, you should clarify "Why does the client want `getServerSideProps` for his site?"

The [main reason](preposterous) for using the `getServerSideProps` is to **fetch data on request time**.

> You should use `getServerSideProps` only if you need to render a page whose data must be fetched at request time.

If that is the case, unfortunately, `getServerSideProps` does not go along with the `react-router-dom` since their 
`Rendering Method` is different as you've noticed.

Although if there is another dirty way around using them together, 
it is inefficient and most likely loses the benefit of both Client Side and Server Side Rendering.

If that is not the case, alternatively you can use `getStaticProps` in order to achieve SEO purposes and so forth.

> The page must be pre-rendered (for SEO) and be very fast — getStaticProps generates HTML and JSON files, both of which can be cached by a CDN for performance

In summary, **No, you cannot use `react-router-dom` and `getServerSideProps` together.**

I feel sorry for your complicated situation and hope you the best of luck to figure out 
the suitable option that fits you and your client's needs.
